File: java/pen/Paper.java
public class Paper {
    private StringBuilder content = new StringBuilder();
    private int symbols;

    public Paper(int symbols) {
        this.symbols = symbols;
    }

    public Paper() {
        this(512);
    }

    public StringBuilder getContent() {
        return this.content;
    }

    public int getSymbols() {
        return this.symbols;
    }

    public void show() {
        System.out.println(this.content.toString());
    }

    public void addContent(String message) throws
        OutOfSpaceException
    {
        if ( this.content.length() == this.symbols ) {
            throw new OutOfSpaceException();
        }
        int available = this.symbols - this.content.length();

        if ( message.length() > available ) {
            this.content.append(message.substring(0, available));
        }
        this.content.append(message);
    }

    public String toString() {
        StringBuilder out = new StringBuilder("Paper(");

        out.append(this.content.length());
        out.append('/');
        out.append(this.symbols);
        out.append(')');

        return out.toString();
    }
}

File: pen/Paper.cpp
#include "Paper.hpp"
#include "exception.hpp"

Paper::Paper(int symbols) {
    this->symbols = symbols;
    this->content = new std::string();
}

Paper::~Paper() {
    delete(this->content);
}

const std::string& Paper::getContent() const {
    return *this->content;
}

int Paper::getSymbols() const {
    return this->symbols;
}

void Paper::addContent(const std::string& message) {
    if ( this->content->length() == this->symbols ) {
        throw OutOfSpaceException();
    }
    int available = this->symbols - this->content->length();

    if ( message.length() > available ) {
        *this->content += message.substr(0, available);
    }
    *this->content += message;
}

void Paper::show() {
    std::cout << *this->content << std::endl;
}


std::ostream& operator<<(std::ostream& out, const Paper& paper) {
    out << "Paper(" << paper.getContent().length() << "/";
    out << paper.getSymbols() << ")";

    return out;
}
